Add Sequencing Metadata to Transaction Context

LogvinovLeon opened the pull request:
## Summary
Adds sequencer-provided metadata (timestamps) to the transaction execution context, making it accessible to modules during transaction processing. This PR ensures type safety throughout the transaction pipeline and guarantees that sequencing metadata flows automatically from transaction acceptance through to storage and API responses.

## Key Changes

### 1. Type System Improvements
- **Moved `FullyBakedTx` to `sov-rollup-interface`**: Relocated from `sov-modules-api` to `rollup-interface/stf` to break a dependency cycle and enable `TransactionReceipt` to store `FullyBakedTx` directly
- **Type-safe transaction storage**: Changed `TransactionReceipt.body_to_save` from `Option<Bytes>` to `Option<FullyBakedTx>`
- **Borsh serialization support**: Changed `FullyBakedTx` fields from `Bytes` to `Vec<u8>` to enable proper serialization with Borsh

### 2. Infrastructure
- **New type**: `HDTimestamp` - u128 nanoseconds since UNIX epoch
- **Extended `FullyBakedTx`**: Added `sequencing_data: Option<Vec<u8>>` field for sequencer metadata
- **Extended `Context`**: Modules access timestamps via `context.sequencing_data()`
- **Helper method**: `FullyBakedTx::set_sequencing_metadata()` for adding metadata
- **Helper method**: `FullyBakedTx::data_as_bytes()` for `Bytes` compatibility

### 3. Sequencer Integration
- Sequencer adds `HDTimestamp::now()` to transactions in side effects handler
- **Automatic metadata flow**: Sequencing data flows automatically from acceptance → execution → `TransactionReceipt` → database → API responses
- Database stores/retrieves full `FullyBakedTx` struct including metadata (serialized with Borsh)
- Metadata survives sequencer restarts and replica synchronization

### 4. Simulation & Testing
- `eth_call` simulation endpoint provides sequencing metadata
- Test framework populates metadata for integration tests
- New integration test validates module access to sequencing data

## Breaking Changes

⚠️ **Database Schema**: Transactions are now stored with full `FullyBakedTx` metadata structure using Borsh serialization. Existing databases contain transactions without the new schema.

**Migration options:**
- Wipe the state and start from 0

⚠️ **API Changes**: `FullyBakedTx.data` and `FullyBakedTx.sequencing_data` are now `Vec<u8>` instead of `Bytes`
- Added `data_as_bytes()` helper method for compatibility

## Testing
✅ Integration test validates module access to sequencing data
✅ Test utilities properly populate metadata

## Technical Details

### Transaction Flow
```
1. Sequencer accepts transaction
   ↓
2. Adds HDTimestamp::now() via set_sequencing_metadata()
   ↓
3. Creates FullyBakedTx with sequencing_data
   ↓
4. Executes transaction → TransactionReceipt
   ↓
5. Receipt.body_to_save contains full FullyBakedTx
   ↓
6. Database serializes FullyBakedTx with Borsh
   ↓
7. API queries return FullyBakedTx with preserved metadata
```

### Code Examples

**Module accessing sequencing data:**
```rust
fn my_module_function<C: Context>(&self, context: &C, ...) {
    if let Some(sequencing_data) = context.sequencing_data() {
        let timestamp: HDTimestamp = borsh::from_slice(&sequencing_data)?;
        // Use timestamp...
    }
}
```

-----

- [ ] I have updated `CHANGELOG.md` with a new entry if my PR makes any breaking changes or fixes a bug. If my PR removes a feature or changes its behavior, I provide help for users on how to migrate to the new behavior.
- [ ] I have carefully reviewed all my `Cargo.toml` changes before opening the PRs. (Are all new dependencies necessary? Is any module dependency leaked into the full-node (hint: it shouldn't)?)

## Linked Issues
- Fixes # (issue, if applicable)
- Related to # (issue) 

## Testing
Describe how these changes were tested. If you've added new features, have you added unit tests?

## Docs
Describe where this code is documented. If it changes a documented interface, have the docs been updated?

Rendered docs are available at `https://sovlabs-ci-rustdoc-artifacts.us-east-1.amazonaws.com/<BRANCH_NAME>/index.html`

<!-- devin-review-badge-begin -->

---

<a href="https://app.devin.ai/review/sovereign-labs/sovereign-sdk/pull/2188">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    <img src="https://static.devin.ai/assets/gh-open-in-devin-review-light.svg?v=1" alt="Open with Devin">
  </picture>
</a>
<!-- devin-review-badge-end -->

LogvinovLeon committed:
Split transaction module into unsigned.rs, v0.rs and v1.rs

LogvinovLeon committed:
Remove redundant custom PartialEq,Eq implementations

LogvinovLeon committed:
Move PubKeyAndSignature into V1

LogvinovLeon committed:
Simplify auth

LogvinovLeon committed:
Add sequencer data to Context

LogvinovLeon committed:
Fix typos

LogvinovLeon committed:
Add sequencing data to FullyBakedTx

LogvinovLeon committed:
Fix clippy

LogvinovLeon committed:
Don't put RawTx but instead FullyBakedTx in Mock Blobs

LogvinovLeon committed:
Fix borsch overhead calculation

LogvinovLeon committed:
Remove postgres test

LogvinovLeon committed:
Regenerate rollup data

LogvinovLeon committed:
Add an empty module for testing sequencing data

LogvinovLeon committed:
Change the discriminant

LogvinovLeon committed:
Implement SequencingDataTester::call

LogvinovLeon committed:
Add integration test for sequencing data

LogvinovLeon committed:
Introduce HDTimestamp and don't pass sequencer metadata separately

LogvinovLeon committed:
Fix Context arguments in tests

LogvinovLeon committed:
Fix sequencing metadata persistence and access

This commit fixes critical bugs where sequencing metadata was being lost:

1. Database persistence: Store and load full FullyBakedTx struct
   - postgres/mod.rs: Serialize entire struct when storing to DB
   - db_data.rs: Deserialize full struct when reading from DB
   - postgres/mod.rs: Deserialize batches correctly

2. Simulation endpoint: Pass metadata to resolve_context()
   - simulate.rs: Create and pass sequencing metadata for eth_call

3. Integration tests: Add metadata to test transactions
   - unregistered_sequencer.rs: Set metadata in test transactions

4. Export HDTimestamp from sov-modules-api for easy access

These changes ensure sequencing metadata (timestamps, etc.) survives
database round-trips and is available to modules in all contexts.

LogvinovLeon committed:
Update CHANGELOG for #2188

LogvinovLeon committed:
Pass None to Context in some test

LogvinovLeon committed:
Cargo fmt

LogvinovLeon committed:
Update lockfiles

LogvinovLeon committed:
Fix a bug in batch_add_tx

LogvinovLeon committed:
Update CHANGELOG

LogvinovLeon committed:
Move FullyBakedTx and RawTx to sov-rollup-interface

Moves FullyBakedTx and RawTx from sov-modules-api to sov-rollup-interface/stf
to establish proper architectural layering. The rollup-interface crate is
the foundational layer for transaction types, and the stf module is where
state transition transaction types belong alongside TransactionReceipt.

Changes:
- Created crates/rollup-interface/src/state_machine/stf/batch.rs with transaction types
- Updated stf/mod.rs to export batch types
- Updated sov-modules-api to re-export from rollup-interface for backward compatibility
- Added module-level documentation

This is a non-breaking change that prepares for making TransactionReceipt.body_to_save
type-safe by storing FullyBakedTx directly instead of raw Bytes.

LogvinovLeon committed:
Make TransactionReceipt.body_to_save type-safe with FullyBakedTx

Changes TransactionReceipt.body_to_save from Option<Bytes> to
Option<FullyBakedTx> to provide type safety and ensure sequencing
metadata flows through the entire transaction pipeline.

This change ensures that:
- FullyBakedTx (including sequencing_data) is stored directly in receipts
- No manual serialization/deserialization needed at construction sites
- The sequencer's timestamp metadata is preserved from acceptance through
  storage to API responses
- Type system prevents storing incorrect data structures

Changes:
- Updated TransactionReceipt.body_to_save type in transaction.rs
- Modified all TransactionReceipt construction sites to use FullyBakedTx
- Removed TODO in block_executor.rs - sequencing_data now flows automatically
- Updated database layer to serialize FullyBakedTx using borsh
- Updated test utilities to construct FullyBakedTx instances

This is a database-breaking change that requires resync.

LogvinovLeon committed:
Run cargo fmt

LogvinovLeon committed:
Change FullyBakedTx to use Vec<u8> instead of Bytes for Borsh compatibility

The bytes::Bytes type does not implement BorshSerialize/BorshDeserialize,
which are required for FullyBakedTx to be serialized for database storage.
This change migrates the data and sequencing_data fields to Vec<u8>.

Changes:
- Changed FullyBakedTx.data from Bytes to Vec<u8>
- Changed FullyBakedTx.sequencing_data from Option<Bytes> to Option<Vec<u8>>
- Added data_as_bytes() helper method for Bytes compatibility
- Updated all call sites to handle Vec<u8> instead of Bytes
- Fixed create_tx_receipt to accept Vec<u8>
- Updated resolve_context call to convert Vec<u8> to Bytes

This maintains the same serialization format while enabling proper Borsh support.

LogvinovLeon committed:
Fix linter

LogvinovLeon committed:
Update test snapshots

LogvinovLeon committed:
Fix linter

bkolad commented on crates/full-node/sov-sequencer/src/preferred/batch_size_tracker.rs:
```diff
@@ -15,8 +15,10 @@ impl BatchSizeTracker {
+    /// BORSH overhead includes:
+    /// - 4 bytes for the Vec length prefix
+    /// - 1 byte for the Option<Bytes> discriminant in FullyBakedTx.sequencing_data
+    const PER_TX_BORSH_OVERHEAD: usize = 5;
```
Isn't it more:?

```
pub struct Bytes {
    ptr: *const u8,
    len: usize,
    // inlined "trait object"
    data: AtomicPtr<()>,
    vtable: &'static Vtable,
}
```

LogvinovLeon replied:
I believe Bytes have custom serialization and are serialized as just a byte array. Same as vec does not serialize it's capacity. Therefore - the only overhead is for Option

bkolad commented on crates/rollup-interface/src/state_machine/stf/batch.rs:
```diff
@@ -0,0 +1,93 @@
+    /// Sequencer-provided metadata for each transaction (e.g., timestamps).
+    /// This data is NOT signed by users but is added by the sequencer.
+    pub sequencing_data: Option<Vec<u8>>,
+}
```
As discussed in the daily, let’s use Bytes.
The following settings should enable the use of Bytes in the rollup interface.
```
borsh = { workspace = true, features = ["bytes"] }
bytes = { workspace = true, features = ["serde"] }
```

LogvinovLeon committed:
Revert "Change FullyBakedTx to use Vec<u8> instead of Bytes for Borsh compatibility"

This reverts commit 2bdbd55407809866babee8d1f41be9b39850673a.

LogvinovLeon committed:
Revert to using Bytes

LogvinovLeon committed:
Fix sequencing data in nonnce queue tests

bkolad commented on crates/full-node/sov-sequencer/src/preferred/transaction_subscriptions.rs:
```diff
@@ -254,6 +254,7 @@ impl<S: Spec, Rt: Runtime<S>> TransactionCache<S, Rt> {
         Ok(Some(AcceptedTx {
             tx: FullyBakedTx {
                 data: tx.body.unwrap_or_default(),
+                sequencing_data: None, // TODO
```
What about this? Is for later, if yes then let's make an issue for it and put it in the comment ()unless it will be done in a followup

bkolad reviewed (commented):
Everything lgtm.
1. Let's add `len()` method to `FullyBakedTx` that will be the data and sequencing_data
2. Let's grep for `baked_tx.data.len()` in the code and see where we should replace it with baked_tx.len()

LogvinovLeon committed:
Store FullyBakedTx instead of Bytes body in more places

LogvinovLeon committed:
Add base64 serialization to FullyBakedTx

Fix deserialization error in WebSocket slot notifications by ensuring
FullyBakedTx serializes both data and sequencing_data fields as base64
strings for JSON compatibility.

Changes:
- Add serde_with with base64 feature to serialize FullyBakedTx fields
- Update OpenAPI spec to define FullyBakedTx schema with base64 fields
- Update LedgerTx.body to reference FullyBakedTx schema

LogvinovLeon committed:
Add len() method to FullyBakedTx and RawTx, replace .data.len() usage

Since FullyBakedTx now contains both data and sequencing_data fields,
using .data.len() no longer gives the correct serialized size. Add a
len() method that returns the total Borsh-serialized length.

Also add len() to RawTx for consistency and convenience.

Updated all callsites that were using .data.len() to use .len() instead,
ensuring correct size calculations for batch size tracking, transaction
size validation, and metrics.

LogvinovLeon committed:
Update snapshots

LogvinovLeon committed:
Update lockfiles

LogvinovLeon committed:
Fix blob tests

LogvinovLeon committed:
Fix last test

LogvinovLeon committed:
Remove file added by accident
